File: ProblemSolving/450DSA/JavaScript/src/matrix/MaximalRectangle.js
/**
 * @author Anirudh Sharma
 * 
 * Given a rows x cols binary matrix filled with 0's and 1's, find the largest rectangle
 * containing only 1's and return its area.
 * 
 * Constraints:
 * 
 * rows == matrix.length
 * cols == matrix.length
 * 0 <= row, cols <= 200
 * matrix[i][j] is '0' or '1'.
 */
const maximalRectangle = (matrix) => {
    // Special cases
    if (matrix === undefined || matrix.length === 0) {
        return 0;
    }
    // Rows and columns of the matrix
    const rows = matrix.length;
    const columns = matrix[0].length;
    // Array to store every row
    const rowArray = Array(columns).fill(0);
    // Fill the row array with the first row of the matrix
    for (let i = 0; i < columns; i++) {
        rowArray[i] = parseInt(matrix[0][i]);
    }
    // Calculate the max area for this row
    let maxArea = largestRectangleInHistogram(rowArray);
    // Loop for every other row
    for (let i = 1; i < rows; i++) {
        // Again fill the row array
        for (let j = 0; j < columns; j++) {
            if (matrix[i][j] === '0') {
                rowArray[j] = 0;
            } else {
                rowArray[j] += parseInt(matrix[i][j]);
            }
        }
        maxArea = Math.max(maxArea, largestRectangleInHistogram(rowArray));
    }
    return maxArea;
};

const largestRectangleInHistogram = (heights) => {
    // Special cases
    if (heights === undefined || heights.length === 0) {
        return 0;
    }
    if (heights.length === 1) {
        return heights[0];
    }
    // Stack to store the index of elements whose value is
    // greater than the top of the stack
    const indices = [];
    // Maximum area of the histogram
    let maxArea = 0;
    // Index to iterate array
    let index = 0;
    // Loop through the copy array
    while (index <= heights.length) {
        if (indices.length === 0 || heights[indices[indices.length - 1]] <= heights[index]) {
            indices.push(index);
            index++;
        } else {
            // Height and width
            let height = heights[indices.pop()];
            let width = indices.length === 0 ? index : index - indices[indices.length - 1] - 1;
            maxArea = Math.max(maxArea, width * height);
        }
    }
    return maxArea;
};

const main = () => {
    let matrix = [["1", "0", "1", "0", "0"], ["1", "0", "1", "1", "1"], ["1", "1", "1", "1", "1"], ["1", "0", "0", "1", "0"]]
    console.log(maximalRectangle(matrix));

    matrix = [];
    console.log(maximalRectangle(matrix));

    matrix = [["0"]];
    console.log(maximalRectangle(matrix));

    matrix = [["1"]];
    console.log(maximalRectangle(matrix));

    matrix = [["1", "1"]];
    console.log(maximalRectangle(matrix));
};

main();

File: ProblemSolving/450DSA/Python/src/matrix/MaximalRectangle.py
"""
@author Anirudh Sharma

Given a rows x cols binary matrix filled with 0's and 1's, find the largest rectangle
containing only 1's and return its area.

Constraints:

rows == matrix.length
cols == matrix.length
0 <= row, cols <= 200
matrix[i][j] is '0' or '1'.
"""


def largestRectangleInHistogram(heights):
    # special cases
    if (heights is None or len(heights) == 0):
        return 0
    if (len(heights) == 1):
        return heights[0]
    # stack to store the index of elements whose value is
    # greater than the top of the stack
    indices = list()
    # Maximum area of the histogram
    maxArea = 0
    # Index to traverse through the heights array
    index = 0
    # Loop through the array
    while index < len(heights):
        if not indices or heights[indices[-1]] <= heights[index]:
            indices.append(index)
            index += 1
        else:
            # Top of the stack
            top = indices.pop()
            # Local area
            localArea = heights[top] * \
                (index - indices[-1] - 1 if indices else index)
            maxArea = max(maxArea, localArea)
    # For remaning elements in the stack
    while indices:
        # Pop the top
        top = indices.pop()
        # Calculate the area with
        # histogram[top_of_stack]
        # stack as smallest bar
        localArea = (heights[top] *
                     ((index - indices[-1] - 1) if indices else index))
        # Update max area, if needed
        maxArea = max(maxArea, localArea)
    return maxArea


def maximalRectangle(matrix):
    # Special cases
    if matrix is None or len(matrix) == 0:
        return 0
    # Rows and columns of the matrix
    rows, columns = len(matrix), len(matrix[0])
    # List to store every row
    rowArray = [None] * columns
    # Fill the rowArray with first row of the matrix
    for i in range(columns):
        rowArray[i] = int(matrix[0][i])
    # Max area for this row
    maxArea = largestRectangleInHistogram(rowArray)
    # Loop for every other row
    for i in range(1, rows):
        # Again fill the row array
        for j in range(columns):
            if matrix[i][j] == "0":
                rowArray[j] = 0
            else:
                rowArray[j] += int(matrix[i][j])
        maxArea = max(maxArea, largestRectangleInHistogram(rowArray))
    return maxArea


if __name__ == "__main__":
    matrix = [["1", "0", "1", "0", "0"], ["1", "0", "1", "1", "1"],
              ["1", "1", "1", "1", "1"], ["1", "0", "0", "1", "0"]]
    print(maximalRectangle(matrix))

    matrix = []
    print(maximalRectangle(matrix))

    matrix = [["0"]]
    print(maximalRectangle(matrix))

    matrix = [["1"]]
    print(maximalRectangle(matrix))

    matrix = [["1", "1"]]
    print(maximalRectangle(matrix))
